prow/cmd/checkconfig/main.go

```diff
@@ -197,10 +198,101 @@ func main() {
 			errs = append(errs, err)
 		}
 	}
+	if o.warningEnabled(tideStrictBranchWarning) {
+		if err := validateStrictBranches(cfg.ProwConfig); err != nil {
+			errs = append(errs, err)
+		}
+	}
 	if len(errs) > 0 {
 		reportWarning(o.strict, errorutil.NewAggregate(errs...))
 	}
 }
+func policyIsStrict(p config.Policy) bool {
+	if p.Protect == nil || !*p.Protect {
+		return false
+	}
+	if p.RequiredStatusChecks == nil || p.RequiredStatusChecks.Strict == nil {
+		return false
+	}
+	return *p.RequiredStatusChecks.Strict
+}
+
+func strictBranchesConfig(c config.ProwConfig) (*orgRepoConfig, error) {
+	strictOrgExceptions := make(map[string]sets.String)
+	strictRepos := sets.NewString()
+	for orgName := range c.BranchProtection.Orgs {
+		org := c.BranchProtection.GetOrg(orgName)
+		// First find explictly configured repos and partition based on strictness.
+		// If any branch in a repo is strict we assume that the whole repo is to
```

Comment:
Might want to emit a warning in case this is not the case, or error out of `checkconfig` just in case someone does do this

Reply:
I don't think we should warn on this and we definitely shouldn't error out. It is totally reasonable to enable strict branch protection on a subset of branches in a repo, this may even be the common case for using strict branch protection.
We would only want to warn if we make this assumption **and** the repo we made the assumption about enables Tide.
I think I could update the code to emit warnings most of the time we make invalid assumptions, but it would add more complexity and still wouldn't detect this case: https://github.com/kubernetes/test-infra/blob/d725f04dad4253751f3996275e5c4ed37a669baa/prow/cmd/checkconfig/main.go#L268-L269

We could instead include a blanket warning that explains how the validation makes assumptions that may be invalid whenever we actually emit an error.
What is your preference?

Reply:
I think best-effort checks for the simple cases we can detect where someone breaks an assumption should be logged at warning level so it's visible to users that something is amiss, at the very least. I'd rather us keep the check code simple, though, so only worth it for areas where we can detect a break of an assumption with minimal added complexity.

Reply:
I suspect that alternative would be very spammy though since we will warn on totally valid and somewhat common cases like I mentioned above. For example, this would imply emitting a warning whenever someone configures branch related query options for Tide (this happens every code freeze).

Reply:
Discussed out of band. Detecting and warning that we have violated assumptions in a way that could produce a false negative will be either excessively spammy or excessively complex.
Going to go with this for now:
> We could instead include a blanket warning that explains how the validation makes assumptions that may be invalid whenever we actually emit an error.